File: BACK/src/main/java/com/genesis/api/statmanager/service/ChampionnatService.java
package com.genesis.api.statmanager.service;

import com.genesis.api.statmanager.dto.championnat.*;
import com.genesis.api.statmanager.dto.global.StatistiquesChampionnatDTO;
import com.genesis.api.statmanager.model.Championnat;
import com.genesis.api.statmanager.model.enumeration.Division;
import java.util.List;

/**
 * 📌 Service pour la gestion des championnats et leurs statistiques.
 */
public interface ChampionnatService {

    /**
     * 📌 Crée un championnat en division donnée avec les joueurs indiqués, puis initialise leurs statistiques.
     *
     * @param division   Division du championnat (DIV1, DIV2, etc.)
     * @param joueursIds Liste des IDs de joueurs à inclure (exactement 23 joueurs).
     * @return ChampionnatDTO créé.
     */
    ChampionnatDTO creerChampionnat(Division division, List<Long> joueursIds);

    /**
     * 📌 Supprime un championnat.
     *
     * @param idChamp ID du championnat à supprimer.
     */
    void supprimerChampionnat(Long idChamp);

    /**
     * 📌 Récupère les 10 derniers championnats.
     *
     * @return Liste de ChampionnatDTO (max 10).
     */
    List<ChampionnatLightDTO> findTop10ByOrderByIdChampDesc(); // ✅ Change le type de retour ici


    void cloturerChampionnat(Long idChampionnat);


    /**
     * 📌 Récupère un championnat avec ses rencontres.
     *
     * @param idChamp ID du championnat.
     * @return ChampionnatDetailWithRencontresDTO.
     */
    ChampionnatDetailWithRencontresDTO findChampionnatWithRencontres(Long idChamp);

    /**
     * 📌 Récupère le récapitulatif (overview) d'un championnat, incluant la liste des derniers championnats.
     *
     * @param idChamp Identifiant du championnat sélectionné.
     * @return ChampionnatOverviewDTO (liste des derniers + détails du sélectionné).
     */
    ChampionnatOverviewDTO getChampionnatOverview(Long idChamp);

    /**
     * 📌 Récupère les statistiques d'un championnat (buts, passes, cotes, etc.).
     *
     * @param idChampionnat L'ID du championnat.
     * @return Liste de StatistiquesChampionnatDTO.
     */
    List<StatistiquesChampionnatDTO> getStatistiquesChampionnat(Long idChampionnat);

    void verifierStatutChampionnat(Championnat championnat);
}

File: front/src/app/modules/championnat/services/championnat.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from './../../../../environments/environment';
import { StatistiquesChampionnatDTO } from './../../statistiques/models/statistiques.model'; // ✅ Import du modèle
import { StatistiquesRencontreDTO } from './../../statistiques/models/statistiques.model';


// Import des DTO que tu utilises
import { ChampionnatOverviewDTO, ChampionnatDetailWithRencontresDTO, ChampionnatDTO, ChampionnatCreationDTO } from './../models/championnat.model'; // par ex. si tu en as un

// ...

@Injectable({
  providedIn: 'root'
})
export class ChampionnatService {
  // ↓ Par exemple : environment.apiBaseUrl = 'http://localhost:8080/api'
  private apiUrl = `${environment.apiBaseUrl}/championnats`; 
  // ou '/championnats' si c'est ton choix

  constructor(private http: HttpClient) {}

  /**
   * Création d'un championnat (POST /championnat/creer)
   * @param division  division ex: 'DIV4'
   * @param joueursIds  liste d'IDs de joueurs (ex: 23 joueurs)
   */
  creerChampionnat(dto: ChampionnatCreationDTO): Observable<ChampionnatDTO> {
    return this.http.post<ChampionnatDTO>(`${this.apiUrl}/create`, dto);
  }

  /**
   * Récupère un "overview" d'un championnat (ou la liste de 10 derniers si pas d'ID)
   * GET /championnat/overview  ou /championnat/overview/{idChamp}
   */
  getChampionnatOverview(idChampionnat?: number): Observable<ChampionnatOverviewDTO> { 
  const url = idChampionnat ? `${this.apiUrl}/overview/${idChampionnat}` : `${this.apiUrl}/overview`;
  return this.http.get<ChampionnatOverviewDTO>(url); // ✅ S'assurer que c'est un objet et pas un tableau
}


  /**
   * Récupère un championnat avec toutes ses rencontres
   * GET /championnat/{idChamp}/details
   */
  getChampionnatWithRencontres(idChamp: number): Observable<ChampionnatDetailWithRencontresDTO> {
    return this.http.get<ChampionnatDetailWithRencontresDTO>(`${this.apiUrl}/${idChamp}/details`);
  }

/**
 * 📌 Récupère les statistiques des joueurs pour une rencontre donnée.
 * GET /rencontres/{idRencontre}/stats
 */
getStatistiquesRencontre(idRencontre: number): Observable<StatistiquesRencontreDTO[]> {
  return this.http.get<StatistiquesRencontreDTO[]>(`${this.apiUrl}/rencontres/${idRencontre}/stats`);
}


/** 📌 Récupère les statistiques des joueurs sur tout un championnat */
getStatistiquesChampionnat(idChampionnat: number): Observable<StatistiquesChampionnatDTO[]> {
  return this.http.get<StatistiquesChampionnatDTO[]>(`${this.apiUrl}/${idChampionnat}/stats`);
}


}
